### PR TITLE
Refactor snmp walk subcommand

### PR DESCRIPTION
**Pull request by dplepage-dd**
## What does this PR do?

This PR refactors the `agent snmp walk` subcommand to fix some issues and make parts of it more reusable.

## Motivation

I am in the process of making another snmp subcommand, and want to reuse most of the arguments and processing used in `snmp walk`; this change will make that much easier. I also found some issues with the snmp walk command, which I have fixed.

### Loading config from the agent

Previously, if the user specified a port number then this subcommand wouldn't try to load any configuration from the agent, requiring the user to provide all relevant data (e.g. authentication data, timeout lengths, etc.). If the user **didn't** provide a port number, then the subcommand would fetch config from the agent, look up the SNMP entry corresponding to the given IP address, and use that data, ignoring any parameters that the user provided.

After this PR, the behavior is that the subcommand will always try to load data from the agent, but any values provided by the user will overwrite those from the agent. If the agent is unreachable, the subcommand will print a warning, but will run anyway (so that it still functions even if you don't have a running agent).

This fixes [NDMII-2450](https://datadoghq.atlassian.net/browse/NDMII-2450) and [NDMII-2541](https://datadoghq.atlassian.net/browse/NDMII-2541)

### Smarter IP parsing

Previously the script assumed that if there was a `:` in the given IP address, then it was a `host:port` pair and the last segment should be parsed as a port. Now, it uses golang's built-in `net.SplitHostPort`, which means that it will correctly identify IPv6 addresses (which also contain colons) and other names; if an IP or hostname contains colons, it must be encased in square brackets, so e.g. to walk localhost on an explicit port you would write
```sh
agent snmp walk '[::1]:161'
```

### Proper version/community defaults

Previously the behavior around defaults was slightly weird - if you didn't pass in a version or passed in an unrecognized version, and also didn't provide a community string, it would complain with a generic "no authentication method provided", but if you explicitly passed in version 1 or 2c then it would default to community="public".

The new behavior is
1. If you pass in an unrecognized version, it complains
2. If you don't pass in any version, but you do provide a username, it defaults to version 3
3. Otherwise, if you don't pass in any version, it defaults to version 2c
4. If it ends up using version 1 or version 2c (regardless of whether you told it to explicitly or it default to it), then a missing community string will default to "public".
5. If it ends up using version 3 and you didn't provide a username, it complains `username is required for snmp v3`.

### Cosmetic changes and bugfixes

 * When a user enters invalid input, the CLI will only show the error message and the usage string, not the full help text (which was obscuring the actual error message previously)
 * Several error messages have been made clearer (in particular, if the local agent is unreachable, it will say this instead of "unexpected end of JSON input")
 * If the datadog config has SNMP entries where network_address is not a valid CIDR address, this subcommand will no longer panic.
 * Counters and Gauges are now formatted as e.g. `Gauge32` instead of `Gauge 32` in the output, to match how `snmpwalk` formats things
 * It will no longer fail to read local config if the local config has `ignored_ip_addresses` set (this is issue NDMII-2541 ).

## Possible Drawbacks / Trade-offs

In the specific case where the agent is reachable but very slow to provide config, it could be frustrating that an snmp walk  with all necessary parameters specified nonetheless checks the config, but this is difficult to avoid without some fairly complex logic to decide how many parameters must be specified before we ignore local config (for example, if you provide all the necessary credentials etc but do not explicitly set a timeout, would you be surprised if the timeout in your config file was ignored?). It is easier to just request it every time, which in nearly cases should be fine - either the agent will reply immediately, or the connection will be refused.

## Describe how to test/QA your changes

Run `agent snmp walk <someaddress>` against the address of a working SNMP agent (you can spin up one on `localhost:161` with `docker run -p 161:161/udp tandrup/snmpsim`) and confirm that it generates the expected data (which should roughly match the output of `snmpwalk -c public <someaddress>:<someport>`).

For thoroughness, I tested it several different ways, including calling it with missing arguments:
 * `agent snmp walk localhost` should default to SNMP v2c with the `public` community on port 161
 * `agent snmp walk localhost:162` should fail immediately because the connection is rejected
 * `agent snmp walk localhost -v 3` should fail immediately, complaining that you must provide a username
 * `agent snmp walk -u simulator -l authPriv -a MD5 -x DES -A auctoritas -X privatus -v 3 --timeout=1 -N 4c9184f37cff01bcdc32dc486ec36961 localhost` should work if you're using `snmpsim` as described above, except I'm not sure how it generates that context name so you might need to look for the context name in the snmpsim logs and replace `4c9184f37cff01bcdc32dc486ec36961` with whatever identifier your copy is using.
 * If you configure your local agent with e.g.
```yaml
snmp_listeners:
  configs:
    - network_address: 127.0.0.1/32
      port: 161
      snmp_version: 3
      timeout: 1
      retries: 3
      user: simulator
      authKey: auctoritas
      authProtocol: MD5
      privKey: privatus
      privProtocol: DES
      context_name: 4c9184f37cff01bcdc32dc486ec36961
```
then `agent snmp walk 127.0.0.1 -v3` should work (again assuming you set the right context name), and overriding any of the auth values should make it stop working (the `-v3` is technically redundant but it makes sure you're not accidentally testing the default `v2c` instead).

[NDMII-2450]: https://datadoghq.atlassian.net/browse/NDMII-2450?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by jmw51798 — APPROVED** [review]
LGTM

File: cmd/agent/subcommands/snmp/command.go [hunk 1]
```diff
@@ -82,219 +75,160 @@ type cliParams struct {
...
+}
+
+// getParams tries to parse deviceAddr as host:port; if it can, it just returns
+// (host, port, params, nil). If the address is not a host:port pair, then it
+// assumes it is a plain IP address, and queries the running agent for the
+// configuration for that address.
+func getParams(deviceAddr string, connParams *connectionParams) (string, uint16, *connectionParams, error) {
+	deviceIP, port, hasPort := parseIPWithMaybePort(deviceAddr)
+
+	if !hasPort {
```
Comment (NouhaManai96) [inline cmd/agent/subcommands/snmp/command.go:111]: this check should be on the number of the arguments and not the port, if we don't have a port ut have other arguments then we should use the default port and the given arguments for the command and not check the config from the agent

Comment (dplepage-dd) [reply cmd/agent/subcommands/snmp/command.go]: I have updated it to always check the config and use the config values for anything that is unset, but use the user-provided values for anything the user provided.

If it can't read config from the agent, or the agent has no config for the given IP, it prints a warning message but continues anyway.

File: cmd/agent/subcommands/snmp/command.go [hunk 2]
```diff
@@ -29,272 +30,267 @@ import (
...
+	snmpCmd := &cobra.Command{
+		Use:   "snmp",
+		Short: "Snmp tools",
+		Long:  ``,
 	}
+
 	snmpWalkCmd := &cobra.Command{
-		Use:   "walk <IP Address>[:Port] [OID] [OPTIONS]",
-		Short: "Perform a snmpwalk, if only a valid IP address is provided with the oid then the agent default snmp config will be used",
+		Use:   "walk <IP Address>[:Port] [OID]",
```
Comment (NouhaManai96) [inline cmd/agent/subcommands/snmp/command.go:85]: why this change since the walk takes other options in addition to the ip address port and OID ? (the original `snmpwalk` command also refers to the rest of the flag as options)

File: cmd/agent/subcommands/snmp/command.go [hunk 3]
```diff
@@ -29,272 +30,267 @@ import (
...
-	defaultCommunityString = "public"
-
-	// snmp v3
-	defaultUserName      = ""
-	defaultAuthProtocol  = ""
-	defaultAuthKey       = ""
-	defaultPrivProtocol  = ""
-	defaultPrivKey       = ""
-	defaultContext       = ""
-	defaultSecurityLevel = ""
```
Comment (NouhaManai96) [inline cmd/agent/subcommands/snmp/command.go:46]: I don't have a strong opinion on this but FYI when we first implemented the command we decided to have these variables even though they all have the same value of an empty string to make it clearer what they represent later on

File: cmd/agent/subcommands/snmp/command.go [hunk 4]
```diff
@@ -29,272 +30,267 @@ import (
...
 		case "aes256":
-			privProtocol = gosnmp.AES256
+			securityParams.PrivacyProtocol = gosnmp.AES256
 		case "aes256c":
-			privProtocol = gosnmp.AES256C
+			securityParams.PrivacyProtocol = gosnmp.AES256C
 		default:
-			fmt.Printf("Unsupported privacy protocol: %s \n", cliParams.privProt)
-			cliParams.cmd.Help() //nolint:errcheck
-			os.Exit(1)
```
Comment (NouhaManai96) [inline cmd/agent/subcommands/snmp/command.go:287]: We actually want to keep this behaviour, if we have unsupported values we want to show the help that lists all the supported values for the different parameters and exit

File: cmd/agent/subcommands/snmp/command.go (hunk 3 above)
Comment (dplepage-dd) [reply cmd/agent/subcommands/snmp/command.go:46]: I dropped them because A) I worried it made it harder to read the lines that were actually setting up the flags, and B) I thought the names were confusing for things like `defaultVersion` and `defaultSecurityLevel`, where the *flags* default to `""` but the actual value that gets used doesn't, because of additional logic that guesses defaults from other arguments. If anyone feels strongly about it I can put these back.

File: cmd/agent/subcommands/snmp/command.go (hunk 4 above)
Comment (dplepage-dd) [reply cmd/agent/subcommands/snmp/command.go:287]: I think printing the full help text makes it hard to spot the actual error message. I've just refactored this again to use a custom Value type for the flags with values so that we get nicely formatted errors from cobra, which I think should solve this problem - now the error message looks like:
```sh
$ agent snmp walk localhost -x foo
Error: invalid argument "foo" for "-x, --priv-protocol" flag: must be one of DES|AES|AES192|AES192C|AES256|AES256C
```

File: cmd/agent/subcommands/snmp/command.go (hunk 2 above)
Comment (dplepage-dd) [reply cmd/agent/subcommands/snmp/command.go:118]: Cobra adds `[flags]` to the end of the help string automatically, so it was showing as
```
Usage:
  agent snmp walk <IP Address>[:Port] [OID] [OPTIONS] [flags]
  ```

File: cmd/agent/subcommands/snmp/command.go (hunk 4 above)
Comment (NouhaManai96) [reply cmd/agent/subcommands/snmp/command.go:287]: much better ! thanks for the change

**Review by NouhaManai96 — APPROVED** [review]
